pil-analyzer/src/type_inference.rs

```diff
@@ -739,46 +723,86 @@ impl TypeChecker {
                 };
                 let struct_decl = self.struct_declarations.get(name).unwrap();
 
-                let scheme = TypeScheme {
-                    ty: Type::TypeVar(struct_decl.name.clone()),
-                    vars: struct_decl.type_vars.clone(),
-                };
-                let (ty, args) = self.unifier.instantiate_scheme(scheme);
+                let fresh_type_vars: Vec<Type> = struct_decl
+                    .type_vars
+                    .vars()
+                    .map(|_| self.unifier.new_type_var())
+                    .collect();
+
                 if let Some(requested_type_args) = type_args {
-                    if requested_type_args.len() != args.len() {
-                        return Err(sr.with_error(format!(
-                            "Expected {} type arguments for struct {name}, but got {}: {}",
-                            args.len(),
-                            requested_type_args.len(),
-                            requested_type_args.iter().join(", ")
-                        )));
-                    }
-                    for (requested, inferred) in requested_type_args.iter_mut().zip(&args) {
-                        requested.substitute_type_vars(&self.declared_type_vars);
+                    for (fresh_var, requested) in
+                        fresh_type_vars.iter().zip(requested_type_args.iter())
+                    {
                         self.unifier
```

Comment:
Why didn't you call the new helper function?

Reply:
The function made it harder to instantiate free type_vars every time, so I reverted the code to its original place and removed the helper function.

Reply:
The code here is way too long with nothing "interesting" going on, please try to combine it with other code.